internal/datacoord/import_scheduler.go

```diff
@@ -0,0 +1,317 @@
+// limitations under the License.
+
+package datacoord
+
+import (
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/samber/lo"
+	"go.uber.org/zap"
+
+	"github.com/milvus-io/milvus/internal/proto/datapb"
+	"github.com/milvus-io/milvus/internal/proto/internalpb"
+	"github.com/milvus-io/milvus/pkg/log"
+	"github.com/milvus-io/milvus/pkg/util/merr"
+)
+
+const (
+	NullNodeID = -1
+)
+
+type ImportScheduler interface {
+	Start()
+	Close()
+}
+
+type importScheduler struct {
+	meta    *meta
+	cluster Cluster
+	alloc   allocator
+	imeta   ImportMeta
+
+	closeOnce sync.Once
+	closeChan chan struct{}
+}
+
+func NewImportScheduler(meta *meta,
+	cluster Cluster,
+	alloc allocator,
+	imeta ImportMeta,
+) ImportScheduler {
+	return &importScheduler{
+		meta:      meta,
+		cluster:   cluster,
+		alloc:     alloc,
+		imeta:     imeta,
+		closeChan: make(chan struct{}),
+	}
+}
+
+func (s *importScheduler) Start() {
+	log.Info("start import scheduler")
+	ticker := time.NewTicker(2 * time.Second)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-s.closeChan:
+			log.Info("import scheduler exited")
+			return
+		case <-ticker.C:
+			s.process()
+		}
+	}
+}
+
+func (s *importScheduler) Close() {
+	s.closeOnce.Do(func() {
+		close(s.closeChan)
+	})
+}
+
+func (s *importScheduler) process() {
+	all := s.imeta.GetBy()
+	tasksByJob := lo.GroupBy(all, func(t ImportTask) int64 {
+		return t.GetJobID()
+	})
+	jobs := lo.Keys(tasksByJob)
+	sort.Slice(jobs, func(i, j int) bool {
+		return jobs[i] < jobs[j]
+	})
+	for _, job := range jobs {
+		tasks := tasksByJob[job]
+		for _, task := range tasks {
+			switch task.GetState() {
+			case internalpb.ImportState_Pending:
+				switch task.GetType() {
+				case PreImportTaskType:
+					s.processPendingPreImport(task)
+				case ImportTaskType:
+					s.processPendingImport(task)
+				}
+			case internalpb.ImportState_InProgress:
+				switch task.GetType() {
+				case PreImportTaskType:
+					s.processInProgressPreImport(task)
+				case ImportTaskType:
+					s.processInProgressImport(task)
+				}
+			case internalpb.ImportState_Completed:
+				s.processCompleted(task)
+			case internalpb.ImportState_Failed:
+				s.processFailed(task)
+			}
+		}
+	}
+}
+
+func (s *importScheduler) checkErr(task ImportTask, err error) {
+	if !merr.IsRetryableErr(err) && !merr.IsCanceledOrTimeout(err) {
+		err = s.imeta.Update(task.GetTaskID(), UpdateState(internalpb.ImportState_Failed), UpdateReason(err.Error()))
+		if err != nil {
+			log.Warn("failed to update import task state to failed", WrapLogFields(task, zap.Error(err))...)
+		}
+		return
+	}
+	err = s.imeta.Update(task.GetTaskID(), UpdateState(internalpb.ImportState_Pending))
+	if err != nil {
+		log.Warn("failed to update import task state to pending", WrapLogFields(task, zap.Error(err))...)
+	}
+}
+
+func (s *importScheduler) getIdleNode() int64 {
+	nodeIDs := lo.Map(s.cluster.GetSessions(), func(s *Session, _ int) int64 {
+		return s.info.NodeID
+	})
+	for _, nodeID := range nodeIDs {
```

Comment:
you can parallel this query to speed up.
refer to 
```
func (nm *IndexNodeManager) PeekClient(meta *model.SegmentIndex) (UniqueID, types.IndexNodeClient) {
//...
}
``` 
in datacoord/indexnode_manager.go